core/dbt/deps/tarball.py

```diff
@@ -0,0 +1,249 @@
+        elif hasattr(fp_in, 'read'):
+            return hashlib.sha1(fp_in.read()).hexdigest()
+        else:
+            raise ValueError()
+
+    def get_tarfile(self):
+        def validate_checksum(self, filepath):
+            if self.sha1:
+                checksum = self.file_sha1(filepath)
+                if not checksum == self.sha1:
+                    package_sha1_fail(checksum, self.sha1)
+
+        def validate_tarfile(self, filepath):
+            if not tarfile.is_tarfile(filepath):
+                msg = f"{filepath} is not a valid tarfile."
+                raise DependencyException(msg)
+
+        def handle_local_tar(self):
+            # pass through local tarfile path primarily for testing
+            validate_checksum(self, self.tarball)
+            validate_tarfile(self, self.tarball)
+            tar = tarfile.open(self.tarball, "r:gz")
+            self.tarfile_bin = tar
+
+        def handle_remote_tar(self):
+            with NamedTemporaryFile(dir=get_downloads_path()) as named_temp_file:
+                download_url = self.tarball
+
+                download_untar_fn = functools.partial(
+                    self.download_tar
+                    , download_url
+                    , named_temp_file.name
+                )
+                connection_exception_retry(download_untar_fn, 5)
+
+                validate_checksum(self, named_temp_file.name)
+                validate_tarfile(self, named_temp_file.name)
+                tar = tarfile.open(named_temp_file.name, "r:gz")
+                self.tarfile_bin = tar
+
+        if type(self.tarball) is str:
+            if os.path.isfile(self.tarball):
+                print('handle_local_tar')
+                handle_local_tar(self)
+            else:
+                # assume download if str and not local file
+                print('handle_remote_tar')
+                handle_remote_tar(self)
+        elif type(self.tarball) is tarfile.TarFile:
+            # pass through TarFile object primarily for testing
+            # no explicit hash test here
+            self.tarfile_bin = self.tarball
+        else:
+            raise ValueError('unknown tarball type defined')
+
+        if not self.get_tar_size() <= TARFILE_MAX_SIZE:
+            msg = ("Tarfile size is larger than limit "
+                   f"of {TARFILE_MAX_SIZE}.")
+            raise DependencyException(msg)
+
+            fire_event(Sha1ChecksumPasses())
+
+    def download_tar(self, download_url, tar_path):
+        """
+        Sometimes the download of the files fails and we want to retry.  Sometimes the
+        download appears successful but the file did not make it through as expected
+        (generally due to a github incident).  Either way we want to retry downloading
+        and untarring to see if we can get a success.  Call this within
+        `_connection_exception_retry`
+        """
+
+        system.download(download_url, tar_path)
```

Comment:
This doesn't need to be a function since all it does it this single line.  Directly call `system.download` instead.

Reply:
makes perfect sense, updated accordingly.